File: src/home/deck_check.c
#include "home/deck_check.h"

#include "generated/wram.h"
#include "home/bg_map.h"
#include "home/random.h"
#include "home/sound.h"
/* >>> factory statics */
#define SYM_SPACE 0x00u
#define SYM_CURSOR_R 0x0Fu

#include "home/deck_check.h"
#include "generated/hram.h"
#include "generated/wram.h"
#include "mem.h"
#define B_PAD_DOWN_MASK 0x80u
#define B_PAD_UP_MASK 0x40u
#define B_PAD_LEFT_MASK 0x20u
#define B_PAD_RIGHT_MASK 0x10u
#define PAD_A_MASK 0x01u
#define PAD_B_MASK 0x02u
#define B_CURSOR_BLINK_PERIOD 4u
#define CURSOR_BLINK_PERIOD_MASK 0x0Fu
#define MENU_CANCEL 0xFFu
#define MENU_CONFIRM 0x01u
#define SFX_CURSOR 0x01u
/* <<< factory statics */

#define SFX_CONFIRM 0x02u
#define SFX_CANCEL 0x03u

DrawCheckMenuCursorResult DrawCheckMenuCursor(uint8_t a)
{
	uint16_t product = HtimesL((uint16_t)((uint16_t)wCheckMenuCursorXPosition << 8 | 10u));
	uint8_t b = (uint8_t)(product + 1u);
	uint8_t c = (uint8_t)((uint8_t)(wCheckMenuCursorYPosition << 1) + 14u);

	WriteByteToBGMap0(a, b, c);
	return (DrawCheckMenuCursorResult){a, a, a == 0 ? 0x80u : 0};
}

void PlaySFXConfirmOrCancel(uint8_t a)
{
	uint8_t sfx_id = (uint8_t)(a + 1u) == 0 ? SFX_CANCEL : SFX_CONFIRM;
	PlaySFX(sfx_id);
}

/* >>> factory EraseCheckMenuCursor */
/* deck_check.asm:95-97 */
DrawCheckMenuCursorResult EraseCheckMenuCursor(void)
{
	return DrawCheckMenuCursor(SYM_SPACE);
}
/* <<< factory EraseCheckMenuCursor */

/* >>> factory DisplayCheckMenuCursor */
/* deck_check.asm:123-125 */
DrawCheckMenuCursorResult DisplayCheckMenuCursor(void)
{
	return DrawCheckMenuCursor(SYM_CURSOR_R);
}
/* <<< factory DisplayCheckMenuCursor */

/* >>> factory HandleCheckMenuInput */
HandleCheckMenuInputResult HandleCheckMenuInput(void)
{
	wMenuInputSFX = 0u;
	uint8_t d = wCheckMenuCursorXPosition;
	uint8_t e = wCheckMenuCursorYPosition;

	uint8_t pad = hDPadHeld;
	if (pad != 0u) {
		uint8_t moved = 0u;
		if ((pad & B_PAD_LEFT_MASK) != 0u) {
			d ^= 0x01u;
			moved = 1u;
		} else if ((pad & B_PAD_RIGHT_MASK) != 0u) {
			d ^= 0x01u;
			moved = 1u;
		} else if ((pad & B_PAD_UP_MASK) != 0u) {
			e ^= 0x01u;
			moved = 1u;
		} else if ((pad & B_PAD_DOWN_MASK) != 0u) {
			e ^= 0x01u;
			moved = 1u;
		}
		if (moved != 0u) {
			wMenuInputSFX = SFX_CURSOR;
			(void)EraseCheckMenuCursor();
			wCheckMenuCursorXPosition = d;
			wCheckMenuCursorYPosition = e;
			wCheckMenuCursorBlinkCounter = 0u;
		}
	}

	uint8_t keys = hKeysPressed;
	uint8_t ab = (uint8_t)(keys & (PAD_A_MASK | PAD_B_MASK));
	if (ab != 0u) {
		if ((ab & PAD_A_MASK) != 0u) {
			(void)DisplayCheckMenuCursor();
			(void)PlaySFXConfirmOrCancel(MENU_CONFIRM);
			return (HandleCheckMenuInputResult){MENU_CONFIRM, 0x10u};
		}
		(void)PlaySFXConfirmOrCancel(MENU_CANCEL);
		return (HandleCheckMenuInputResult){MENU_CANCEL, 0x90u};
	}

	if (wMenuInputSFX != 0u)
		PlaySFX(wMenuInputSFX);

	uint8_t old_counter = wCheckMenuCursorBlinkCounter;
	wCheckMenuCursorBlinkCounter = (uint8_t)(old_counter + 1u);
	uint8_t masked = (uint8_t)(old_counter & CURSOR_BLINK_PERIOD_MASK);
	if (masked != 0u)
		return (HandleCheckMenuInputResult){masked, 0x20u};

	if ((wCheckMenuCursorBlinkCounter & (1u << B_CURSOR_BLINK_PERIOD)) == 0u) {
		DrawCheckMenuCursorResult r = DrawCheckMenuCursor(SYM_CURSOR_R);
		return (HandleCheckMenuInputResult){r.a, r.f};
	}
	DrawCheckMenuCursorResult r2 = EraseCheckMenuCursor();
	return (HandleCheckMenuInputResult){r2.a, r2.f};
}
/* <<< factory HandleCheckMenuInput */

File: tests/cases/deck_check.py
POISON = {"a": 0xAA, "f": 0xF0, "b": 0xBB, "c": 0xCC,
          "d": 0xDD, "e": 0xEE, "hl": 0x1234}

X_POSITION = 0xCEAF
Y_POSITION = 0xCEB0

CONTRACT = {
    "DrawCheckMenuCursor": {
        "compare": ("a", "d", "e", "f"),
        "preserve": ("d",),
    },
    "PlaySFXConfirmOrCancel": {
        "compare": ("a", "f", "b", "c", "d", "e", "hl"),
        "preserve": ("a", "f", "b", "c", "d", "e", "hl"),
    },
}

CASES = {
    "DrawCheckMenuCursor": [
        {"read": {0x9800 + 14 * 32 + 1: 1}},
        dict(POISON, a=0x5A, wram={X_POSITION: b"\x01", Y_POSITION: b"\x01"},
             read={0x9800 + 16 * 32 + 11: 1}),
        {"a": 0x33, "wram": {X_POSITION: b"\xFF", Y_POSITION: b"\xFF"},
         "read": {0x9800 + 12 * 32 + 247: 1}},
    ],
    "PlaySFXConfirmOrCancel": [
        {"a": 0, "read": {0xDD82: 1, 0xDD83: 1}},
        dict(POISON, a=0xFF, read={0xDD82: 1, 0xDD83: 1}),
        {"a": 1, "read": {0xDD82: 1, 0xDD83: 1}},
    ],
}

# >>> factory EraseCheckMenuCursor
CONTRACT["EraseCheckMenuCursor"] = {"compare": ("a", "d", "e", "f"), "preserve": ("d",)}
CASES["EraseCheckMenuCursor"] = [
    {"read": {0x9800 + 14 * 32 + 1: 1}},
    dict(POISON, wram={X_POSITION: b"\x01", Y_POSITION: b"\x01"},
         read={0x9800 + 16 * 32 + 11: 1}),
    {"wram": {X_POSITION: b"\xFF", Y_POSITION: b"\xFF"},
     "read": {0x9800 + 12 * 32 + 247: 1}},
]
# <<< factory EraseCheckMenuCursor

# >>> factory DisplayCheckMenuCursor
CONTRACT["DisplayCheckMenuCursor"] = {"compare": ("a", "d", "e", "f"), "preserve": ("d",)}
CASES["DisplayCheckMenuCursor"] = [
    {"read": {0x9800 + 14 * 32 + 1: 1}},
    dict(POISON, wram={X_POSITION: b"\x01", Y_POSITION: b"\x01"},
         read={0x9800 + 16 * 32 + 11: 1}),
    {"wram": {X_POSITION: b"\xFF", Y_POSITION: b"\xFF"},
     "read": {0x9800 + 12 * 32 + 247: 1}},
]
# <<< factory DisplayCheckMenuCursor

# >>> factory-cases-statics
wMenuInputSFX = 0xCFE3
hDPadHeld = 0xFF8F
hKeysPressed = 0xFF91
wCheckMenuCursorXPosition = 0xCEAF
wCheckMenuCursorYPosition = 0xCEB0
wCheckMenuCursorBlinkCounter = 0xCEA3
# <<< factory-cases-statics

# >>> factory HandleCheckMenuInput
CONTRACT["HandleCheckMenuInput"] = {"compare": ("a", "f"), "preserve": (), "wram_out": True}
CASES["HandleCheckMenuInput"] = [
    {"wram": {hDPadHeld: b"\x00", hKeysPressed: b"\x02",
              wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03"}},
    dict(POISON, wram={hDPadHeld: b"\x00", hKeysPressed: b"\x03",
                       wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03"}),
    {"wram": {hDPadHeld: b"\x20", hKeysPressed: b"\x00",
              wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03",
              wCheckMenuCursorBlinkCounter: b"\x01"}},
    {"wram": {hDPadHeld: b"\x00", hKeysPressed: b"\x00",
              wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03",
              wCheckMenuCursorBlinkCounter: b"\x05"}},
    {"wram": {hDPadHeld: b"\x00", hKeysPressed: b"\x00",
              wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03",
              wCheckMenuCursorBlinkCounter: b"\x00"}},
    {"wram": {hDPadHeld: b"\x00", hKeysPressed: b"\x00",
              wCheckMenuCursorXPosition: b"\x05", wCheckMenuCursorYPosition: b"\x03",
              wCheckMenuCursorBlinkCounter: b"\x10"}},
]
# <<< factory HandleCheckMenuInput

from tests.cases._schema_migration import legacy_to_schema
SCHEMA2_CASES = legacy_to_schema(CASES, CONTRACT)

MUTATIONS = {
    "DrawCheckMenuCursor": {
        "source_symbol": "DrawCheckMenuCursor",
        "before": "uint8_t b = (uint8_t)(product + 1u);",
        "after": "uint8_t b = (uint8_t)(product + 2u);",
        "case_ids": ["DrawCheckMenuCursor-1", "DrawCheckMenuCursor-2"],
    },
    "PlaySFXConfirmOrCancel": {
        "source_symbol": "PlaySFXConfirmOrCancel",
        "before": "uint8_t sfx_id = (uint8_t)(a + 1u) == 0 ? SFX_CANCEL : SFX_CONFIRM;",
        "after": "uint8_t sfx_id = (uint8_t)(a + 1u) == 0 ? SFX_CONFIRM : SFX_CONFIRM;",
        "case_ids": ["PlaySFXConfirmOrCancel-1"],
    },
}
# >>> factory-mutation EraseCheckMenuCursor
MUTATIONS["EraseCheckMenuCursor"] = {
    "source_symbol": "EraseCheckMenuCursor",
    "before": "DrawCheckMenuCursorResult EraseCheckMenuCursor(void)\n{\n\treturn DrawCheckMenuCursor(SYM_SPACE);\n}",
    "after": "DrawCheckMenuCursorResult EraseCheckMenuCursor(void)\n{\n\treturn DrawCheckMenuCursor(SYM_CURSOR_R);\n}",
    "case_ids": ["EraseCheckMenuCursor-0", "EraseCheckMenuCursor-1", "EraseCheckMenuCursor-2"],
}
# <<< factory-mutation EraseCheckMenuCursor
# >>> factory-mutation DisplayCheckMenuCursor
MUTATIONS["DisplayCheckMenuCursor"] = {
    "source_symbol": "DisplayCheckMenuCursor",
    "before": "DrawCheckMenuCursorResult DisplayCheckMenuCursor(void)\n{\n\treturn DrawCheckMenuCursor(SYM_CURSOR_R);\n}",
    "after": "DrawCheckMenuCursorResult DisplayCheckMenuCursor(void)\n{\n\treturn DrawCheckMenuCursor(SYM_SPACE);\n}",
    "case_ids": ["DisplayCheckMenuCursor-0", "DisplayCheckMenuCursor-1", "DisplayCheckMenuCursor-2"],
}
# <<< factory-mutation DisplayCheckMenuCursor
# >>> factory-mutation HandleCheckMenuInput
MUTATIONS["HandleCheckMenuInput"] = {"source_symbol": "HandleCheckMenuInput", "before": "return (HandleCheckMenuInputResult){MENU_CONFIRM, 0x10u};", "after": "return (HandleCheckMenuInputResult){MENU_CANCEL, 0x10u};", "case_ids": ["HandleCheckMenuInput-1"]}
# <<< factory-mutation HandleCheckMenuInput
